File: exercism/python/grade-school/school.py
from collections import defaultdict


class School(object):
    def __init__(self, name):
        self._store = defaultdict(list)

    def add(self, student, grade):
        self._store[grade].append(student)

    def grade(self, g):
        return tuple(sorted(self._store[g]))

    def sort(self):
        return sorted((key, self.grade(key)) for key in self._store)

File: exercism/haskell/grade-school/School.hs
module School (School, add, empty, sorted, grade) where

import Data.List (sort)
import qualified Data.Map as Map

type Grade = Int
type Student = String
type School = Map.Map Grade [Student]

empty :: School
empty = Map.empty

add :: Grade -> Student -> School -> School
add k v = Map.insertWith (++) k [v]

sorted :: School -> [(Grade, [Student])]
sorted = Map.toList . Map.map sort

grade :: Grade -> School -> [Student]
grade = Map.findWithDefault []
